### PR TITLE
fix: add duplicate key detection for values.yaml

### LINKED ISSUE #31102: Helm allows installing a chart using invalid YAML in values.yaml
Helm allows installing a chart using invalid YAML in values.yaml

**Reproduction steps**
- Create a helm chart using `helm create foo`
- Inject invalid YAML in the file
```sh
echo -e "invalid:
  duplicate: default
  duplicate: value-i-want
  duplicate: last-one
" >> foo/values.yaml
```
- Install the package using `helm install foo foo`
- Check values of installed release
```sh
helm get  values foo --all | grep invalid -A3
invalid:
  duplicate: last-one
livenessProbe:
```

Note that some of the values are removed leaving only one. It looks like helm attempts to correct the yaml by removing duplicates.

### PR DESCRIPTION
**Pull request by KartikSuryavanshi**
## Description

This PR adds a lint rule to detect duplicate keys in . This addresses the issue where Helm allows installing charts with invalid YAML that contains duplicate keys, which silently takes the last value.

## Changes

- Add  and  functions in  that use 
- Add  function in both v2 and v3 lint rules to check for duplicate keys
- Add test data with duplicate keys for both v2 and v3
- Add tests to verify duplicate key detection works

## Testing

The new lint rule will detect duplicate keys in  and report an error. For example:

Will produce an error like:

## Related Issues

Closes #31102

### CODE REVIEW

**Comment by promptless-for-oss** [comment]
Promptless prepared a documentation update related to this change.

Triggered by [helm/helm PR #32363](https://github.com/helm/helm/pull/32363)

Added an "Avoid Duplicate Keys" section to the Chart Best Practices → Values guide, explaining that duplicate keys in `values.yaml` are invalid YAML (Helm silently keeps the last value) and that `helm lint` now reports them as an error. Documented as a lint-time check, since the strict duplicate-key detection is wired into `helm lint`.

Review: [docs: document duplicate-key detection in values.yaml](https://github.com/helm/helm-www/pull/2181)

File: internal/sympath/walk.go [hunk 1]
```diff
@@ -68,7 +67,14 @@ func symwalk(path string, info os.FileInfo, walkFn filepath.WalkFunc) error {
 	if IsSymlink(info) {
 		resolved, err := filepath.EvalSymlinks(path)
 		if err != nil {
-			return fmt.Errorf("error evaluating symlink %s: %w", path, err)
+			if os.IsNotExist(err) {
+				// If a symlink's target does not exist (broken symlink),
+				// silently skip it. Broken symlinks do not contribute
+				// content and should not cause errors, especially when
```
[the remark this answers is not in the document]
Comment (KartikSuryavanshi) [reply internal/sympath/walk.go]: done

File: internal/sympath/walk.go [hunk 2]
```diff
@@ -68,7 +67,14 @@ func symwalk(path string, info os.FileInfo, walkFn filepath.WalkFunc) error {
...
+				// they match .helmignore patterns.
+				return nil
```
[the remark this answers is not in the document]
Comment (KartikSuryavanshi) [reply internal/sympath/walk.go]: done

File: pkg/chart/common/values_test.go [hunk 3]
```diff
@@ -203,3 +204,34 @@ chapter:
...
+  duplicate: last-one
+`
+
+	_, err := ReadValuesStrict([]byte(doc))
+	if err == nil {
+		t.Fatal("Expected error for duplicate keys, got nil")
+	}
+	if !strings.Contains(err.Error(), "already set in map") {
+		t.Fatalf("Expected duplicate-key error, got: %s", err)
+	}
```
[the remark this answers is not in the document]
Comment (KartikSuryavanshi) [reply pkg/chart/common/values_test.go]: done

File: internal/sympath/walk.go [hunk 4]
```diff
@@ -68,7 +67,14 @@ func symwalk(path string, info os.FileInfo, walkFn filepath.WalkFunc) error {
...
+			}
```
[the remark this answers is not in the document]
Comment (KartikSuryavanshi) [reply internal/sympath/walk.go]: done

File: internal/sympath/walk.go [hunk 5]
```diff
@@ -68,7 +67,14 @@ func symwalk(path string, info os.FileInfo, walkFn filepath.WalkFunc) error {
...
+			return err
```
[the remark this answers is not in the document]
Comment (KartikSuryavanshi) [reply internal/sympath/walk.go]: done

File: internal/sympath/walk.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (KartikSuryavanshi) [reply internal/sympath/walk.go]: done

File: internal/chart/v3/lint/lint_test.go [hunk 6]
```diff
@@ -241,3 +242,25 @@ func TestMalformedTemplate(t *testing.T) {
...
+		t.Fatalf("All didn't fail with expected errors, got %#v", m)
+	}
+	// Find the duplicate keys error message
+	found := false
+	for _, msg := range m {
+		if strings.Contains(msg.Err.Error(), "failed to strictly parse values file") {
+			found = true
+			break
+		}
+	}
```
[the remark this answers is not in the document]
Comment (KartikSuryavanshi) [reply internal/chart/v3/lint/lint_test.go:263]: done

File: pkg/chart/common/values_test.go [hunk 7]
```diff
@@ -203,3 +203,40 @@ chapter:
...
+	if !contains(err.Error(), "duplicate key") && !contains(err.Error(), "already defined") {
+		// The error message may vary depending on the YAML library version
+		// but it should indicate a duplicate key issue
+		t.Logf("Got error as expected: %s", err)
+	}
+}
+
+func contains(s, substr string) bool {
+	return len(s) >= len(substr) && (s == substr || len(s) > 0 && (s[0:len(substr)] == substr || contains(s[1:], substr)))
+}
```
[the remark this answers is not in the document]
Comment (KartikSuryavanshi) [reply pkg/chart/common/values_test.go]: done

File: pkg/chart/v2/lint/lint_test.go [hunk 8]
```diff
@@ -245,3 +246,25 @@ func TestMalformedTemplate(t *testing.T) {
...
+		t.Fatalf("All didn't fail with expected errors, got %#v", m)
+	}
+	// Find the duplicate keys error message
+	found := false
+	for _, msg := range m {
+		if strings.Contains(msg.Err.Error(), "failed to strictly parse values file") {
+			found = true
+			break
+		}
+	}
```
[the remark this answers is not in the document]
Comment (KartikSuryavanshi) [reply pkg/chart/v2/lint/lint_test.go:267]: done

File: pkg/chart/common/values_test.go [hunk 9]
```diff
@@ -203,3 +203,40 @@ chapter:
...
+
+	_, err := ReadValuesStrict([]byte(doc))
+	if err == nil {
+		t.Fatal("Expected error for duplicate keys, got nil")
+	}
+	if !contains(err.Error(), "duplicate key") && !contains(err.Error(), "already defined") {
+		// The error message may vary depending on the YAML library version
+		// but it should indicate a duplicate key issue
+		t.Logf("Got error as expected: %s", err)
+	}
```
[the remark this answers is not in the document]
Comment (KartikSuryavanshi) [reply pkg/chart/common/values_test.go]: done

File: pkg/chart/v2/lint/rules/values.go [hunk 10]
```diff
@@ -83,3 +88,11 @@ func validateValuesFile(valuesPath string, overrides map[string]any, skipSchemaV
...
 	return nil
 }
+
+func validateValuesFileDuplicateKeys(valuesPath string) error {
+	_, err := common.ReadValuesFileStrict(valuesPath)
+	if err != nil {
+		return fmt.Errorf("failed to strictly parse values file: %w", err)
+	}
+	return nil
+}
```
[the remark this answers is not in the document]
Comment (KartikSuryavanshi) [reply pkg/chart/v2/lint/rules/values.go]: done

File: internal/chart/v3/lint/rules/values.go [hunk 11]
```diff
@@ -83,3 +88,11 @@ func validateValuesFile(valuesPath string, overrides map[string]any, skipSchemaV
...
 	return nil
 }
+
+func validateValuesFileDuplicateKeys(valuesPath string) error {
+	_, err := common.ReadValuesFileStrict(valuesPath)
+	if err != nil {
+		return fmt.Errorf("failed to strictly parse values file: %w", err)
+	}
+	return nil
+}
```
[the remark this answers is not in the document]
Comment (KartikSuryavanshi) [reply internal/chart/v3/lint/rules/values.go]: done

File: pkg/chart/common/values_test.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (KartikSuryavanshi) [reply pkg/chart/common/values_test.go]: done

File: pkg/chart/v2/lint/lint_test.go [hunk 12]
```diff
@@ -245,3 +246,26 @@ func TestMalformedTemplate(t *testing.T) {
...
+	// Expect at least 1 message (the duplicate keys error)
+	if len(m) < 1 {
+		t.Fatalf("All didn't fail with expected errors, got %#v", m)
+	}
+	// Find the duplicate keys error message
+	found := false
+	for _, msg := range m {
+		if msg.Path == "values.yaml" && msg.Severity == support.ErrorSev &&
+			strings.Contains(msg.Err.Error(), "failed to strictly parse values file") {
+			found = true
```
[the remark this answers is not in the document]
Comment (KartikSuryavanshi) [reply pkg/chart/v2/lint/lint_test.go:264]: done

File: internal/chart/v3/lint/lint_test.go [hunk 13]
```diff
@@ -241,3 +242,26 @@ func TestMalformedTemplate(t *testing.T) {
...
+	// Expect at least 1 message (the duplicate keys error)
+	if len(m) < 1 {
+		t.Fatalf("All didn't fail with expected errors, got %#v", m)
+	}
+	// Find the duplicate keys error message
+	found := false
+	for _, msg := range m {
+		if msg.Path == "values.yaml" && msg.Severity == support.ErrorSev &&
+			strings.Contains(msg.Err.Error(), "failed to strictly parse values file") {
+			found = true
```
[the remark this answers is not in the document]
Comment (KartikSuryavanshi) [reply internal/chart/v3/lint/lint_test.go:260]: done

File: internal/chart/v3/lint/rules/values.go [hunk 14]
```diff
@@ -59,6 +59,11 @@ func validateValuesFile(valuesPath string, overrides map[string]any, skipSchemaV
 		return fmt.Errorf("unable to parse YAML: %w", err)
 	}
 
+	// Check for duplicate keys in values.yaml
+	if err := validateValuesFileDuplicateKeys(valuesPath); err != nil {
+		return err
```
[the remark this answers is not in the document]
Comment (KartikSuryavanshi) [reply internal/chart/v3/lint/rules/values.go]: done

File: pkg/chart/v2/lint/rules/values.go [hunk 15]
```diff
@@ -59,6 +59,11 @@ func validateValuesFile(valuesPath string, overrides map[string]any, skipSchemaV
 		return fmt.Errorf("unable to parse YAML: %w", err)
 	}
 
+	// Check for duplicate keys in values.yaml
+	if err := validateValuesFileDuplicateKeys(valuesPath); err != nil {
+		return err
```
[the remark this answers is not in the document]
Comment (KartikSuryavanshi) [reply pkg/chart/v2/lint/rules/values.go]: done

File: pkg/chart/v2/lint/rules/values.go [hunk 16]
```diff
@@ -83,3 +88,10 @@ func validateValuesFile(valuesPath string, overrides map[string]any, skipSchemaV
 
 	return nil
 }
+
+func validateValuesFileDuplicateKeys(valuesPath string) error {
+	if _, err := common.ReadValuesFileStrict(valuesPath); err != nil {
+		return fmt.Errorf("%s contains duplicate keys: %w", filepath.Base(valuesPath), err)
+	}
```
[the remark this answers is not in the document]
Comment (KartikSuryavanshi) [reply pkg/chart/v2/lint/rules/values.go]: done

File: internal/chart/v3/lint/rules/values.go [hunk 17]
```diff
@@ -83,3 +88,10 @@ func validateValuesFile(valuesPath string, overrides map[string]any, skipSchemaV
 
 	return nil
 }
+
+func validateValuesFileDuplicateKeys(valuesPath string) error {
+	if _, err := common.ReadValuesFileStrict(valuesPath); err != nil {
+		return fmt.Errorf("%s contains duplicate keys: %w", filepath.Base(valuesPath), err)
+	}
```
[the remark this answers is not in the document]
Comment (KartikSuryavanshi) [reply internal/chart/v3/lint/rules/values.go]: done

File: pkg/chart/common/values_test.go [hunk 18]
```diff
@@ -203,3 +204,34 @@ chapter:
...
+poet: "Coleridge"
+title: "Rime of the Ancient Mariner"
+`
+
+	data, err := ReadValuesStrict([]byte(doc))
+	if err != nil {
+		t.Fatalf("Error parsing bytes: %s", err)
+	}
+	if data["poet"] != "Coleridge" {
+		t.Errorf("Unexpected poet: %s", data["poet"])
```
[the remark this answers is not in the document]
Comment (KartikSuryavanshi) [reply pkg/chart/common/values_test.go]: done

File: pkg/chart/common/values.go [hunk 19]
```diff
@@ -118,6 +118,26 @@ func ReadValuesFile(filename string) (Values, error) {
 	return ReadValues(data)
 }
 
+// ReadValuesFileStrict will parse a YAML file into a map of values using strict unmarshaling.
+// This will detect duplicate keys in the YAML.
+func ReadValuesFileStrict(filename string) (Values, error) {
```
[the remark this answers is not in the document]
Comment (KartikSuryavanshi) [reply pkg/chart/common/values.go:123]: done

File: internal/chart/v3/lint/rules/values.go [hunk 20]
```diff
@@ -54,9 +55,20 @@ func validateValuesFileExistence(valuesPath string) error {
...
+			return fmt.Errorf("%s contains duplicate keys: %w", filepath.Base(valuesPath), strictErr)
+		}
+		// Fall back to non-strict parsing for general YAML errors
+		_, parseErr := common.ReadValuesFile(valuesPath)
+		if parseErr != nil {
+			return fmt.Errorf("unable to parse YAML: %w", parseErr)
+		}
+		// If non-strict parsing succeeded but strict failed, it's still a duplicate key issue
+		return fmt.Errorf("%s contains duplicate keys: %w", filepath.Base(valuesPath), strictErr)
 	}
```
[the remark this answers is not in the document]
Comment (KartikSuryavanshi) [reply internal/chart/v3/lint/rules/values.go:67]: done

File: pkg/chart/v2/lint/rules/values.go [hunk 21]
```diff
@@ -54,9 +55,20 @@ func validateValuesFileExistence(valuesPath string) error {
...
+			return fmt.Errorf("%s contains duplicate keys: %w", filepath.Base(valuesPath), strictErr)
+		}
+		// Fall back to non-strict parsing for general YAML errors
+		_, parseErr := common.ReadValuesFile(valuesPath)
+		if parseErr != nil {
+			return fmt.Errorf("unable to parse YAML: %w", parseErr)
+		}
+		// If non-strict parsing succeeded but strict failed, it's still a duplicate key issue
+		return fmt.Errorf("%s contains duplicate keys: %w", filepath.Base(valuesPath), strictErr)
 	}
```
[the remark this answers is not in the document]
Comment (KartikSuryavanshi) [reply pkg/chart/v2/lint/rules/values.go:65]: done

File: pkg/chart/v2/lint/rules/values.go [hunk 22]
```diff
@@ -54,9 +55,13 @@ func validateValuesFileExistence(valuesPath string) error {
...
-	if err != nil {
-		return fmt.Errorf("unable to parse YAML: %w", err)
+	// Try strict parsing first to detect duplicate keys
+	values, strictErr := common.ReadValuesFileStrict(valuesPath)
+	if strictErr != nil {
+		if isDuplicateKeyError(strictErr) {
+			return fmt.Errorf("%s contains duplicate keys: %w", filepath.Base(valuesPath), strictErr)
+		}
+		return fmt.Errorf("unable to parse YAML: %w", strictErr)
 	}
```
[the remark this answers is not in the document]
Comment (KartikSuryavanshi) [reply pkg/chart/v2/lint/rules/values.go:65]: This is lint-only by design. The issue discussion (particularly @gjenkins8's comment) recommended starting with a lint rule as the first step. Extending strict detection into the loader/install path would be a separate PR with different backward-compatibility implications, as charts that are malformed but still install would break. See the issue thread for context.

File: internal/chart/v3/lint/rules/values.go [hunk 23]
```diff
@@ -83,3 +88,11 @@ func validateValuesFile(valuesPath string, overrides map[string]any, skipSchemaV
 
 	return nil
 }
+
+// isDuplicateKeyError checks if an error is related to duplicate YAML keys
+func isDuplicateKeyError(err error) bool {
+	errStr := err.Error()
```
Comment (TerryHowe) [inline internal/chart/v3/lint/rules/values.go]: Is there error wrapped so we can't get the type and have to use the string. Do the test cases cover each one of these 3 scenarios because it is kind of surprising there would be so many variations.

Comment (KartikSuryavanshi) [reply internal/chart/v3/lint/rules/values.go]: @TerryHowe Good catch on both points. The error IS wrapped (fmt.Errorf with %w), but errors.As properly unwraps it. I've replaced the string matching with a type-based check:

func isDuplicateKeyError(err error) bool {
    var typeErr *yamlv2.TypeError
    return errors.As(err, &typeErr)
}

The go.yaml.in/yaml/v2 source confirms it only ever produces "already set in map" for duplicate keys — the other two strings ("already defined" and "duplicate") were dead code. The type-based approach eliminates this entirely and is more robust.

I've also added tests covering: the direct error case, the wrapped error case (which is what sigs.k8s.io/yaml produces), non-duplicate errors, nil, and an integration test with actual duplicate YAML.

File: internal/chart/v3/lint/rules/values_test.go [hunk 24]
```diff
@@ -181,3 +184,53 @@ func createTestingSchema(t *testing.T, dir string) string {
...
+			name:     "non-duplicate key error",
+			err:      errors.New("some other error"),
+			expected: false,
+		},
+		{
+			name:     "nil error",
+			err:      nil,
+			expected: false,
+		},
+	}
```
[the remark this answers is not in the document]
Comment (KartikSuryavanshi) [reply internal/chart/v3/lint/rules/values_test.go:219]: done

File: internal/chart/v3/lint/rules/values.go [hunk 25]
```diff
@@ -83,3 +89,9 @@ func validateValuesFile(valuesPath string, overrides map[string]any, skipSchemaV
 
 	return nil
 }
+
+// isDuplicateKeyError checks if an error is related to duplicate YAML keys
+func isDuplicateKeyError(err error) bool {
+	var typeErr *yamlv2.TypeError
+	return errors.As(err, &typeErr)
+}
```
[the remark this answers is not in the document]
Comment (KartikSuryavanshi) [reply internal/chart/v3/lint/rules/values.go]: done
